planner/core/rule_aggregation_elimination.go

```diff
@@ -86,28 +92,49 @@ func (a *aggregationEliminateChecker) tryToEliminateDistinct(agg *LogicalAggrega
 			if canEliminate {
 				distinctByUniqueKey := false
 				schemaByDistinct := expression.NewSchema(cols...)
+				var uniqueKey expression.KeyInfo
 				for _, key := range agg.children[0].Schema().Keys {
 					if schemaByDistinct.ColumnsIndices(key) != nil {
 						distinctByUniqueKey = true
+						uniqueKey = key
 						break
 					}
 				}
 				for _, key := range agg.children[0].Schema().UniqueKeys {
 					if schemaByDistinct.ColumnsIndices(key) != nil {
 						distinctByUniqueKey = true
+						uniqueKey = key
 						break
 					}
 				}
 				if distinctByUniqueKey {
 					af.HasDistinct = false
-					// TODO: fulfill in future pr
-					opt.appendStepToCurrent(agg.ID(), agg.TP(), "", "")
+					appendDistinctEliminateTraceStep(agg, uniqueKey, af, opt)
 				}
 			}
 		}
 	}
 }
 
+func appendAggregationEliminateTraceStep(agg *LogicalAggregation, uniqueKey expression.KeyInfo, opt *logicalOptimizeOp) {
+	if opt.tracer == nil {
```

Comment:
Why do you check null here again, since you've already wrap tracer to logicalOptimizeOp and do check in logicalOptimizeOp? You can simplify here. Since wrap, rule should not see and call tracer directly.

Reply:
And after simplification, you do not need the 2 funcs here, just call appendStep... in tryToEliminate.

Reply:
updated, I tend to wrap `appendStep` in an extra function due to sometimes constructing `reason` and `action` will be complex, thus I separate it from the main logic.